docs/domains/sparse_linear_algebra.rst

```diff
@@ -31,3 +31,137 @@ Currently known limitations:
   ``oneapi::mkl::unimplemented`` exception.
 - Scalar parameters ``alpha`` and ``beta`` should be host pointers to prevent
   synchronizations and copies to the host.
+
+
+cuSPARSE backend
+----------------
+
+Currently known limitations:
+
+- Using ``spmv`` with a ``type_view`` other than ``matrix_descr::general`` will
+  throw an ``oneapi::mkl::unimplemented`` exception.
+- The COO format require the indices to be sorted by row.
```

Comment:
Highlighting this limitation. So far the tests were shuffling all the indices of the COO format. Now the tests check the device type and only shuffle the indices within a row for Nvidia GPU (and will likely need to do the same for AMD). This is documented in the cuSPARSE doc here: https://docs.nvidia.com/cuda/cusparse/index.html#coordinate-coo
```quote
The COO format is assumed to be sorted by row
```

I was wondering if we should introduce a matrix property for COO format sorted by row but that would mean the user would always need to set that property to use the cuSPARSE backend which seems very error prone. We could automatically set that property but as of today that's equivalent to not having the property at all. Let me know your thoughts.

Reply:
This is a gap for which I don't have a clean solution yet. COO, unfortunately, has multiple states of sortedness:
* Unsorted <- nonunique
* Sorted by row, but unsorted column indices within each row <- nonunique
* Sorted by rows then by columns (CSR style) <- unique
* Sorted by column, but unsorted row indices within each column <- nonunique
* Sorted by columns then by rows (CSC style) <- unique

Even though it is easy to transpose a COO matrix by just interchanging the row and column index arrays, we still cannot eliminate the last 2 pointers above as being handled by 2nd and 3rd pointers because not all cases are covered (e.g., `complex` + `conjtrans`). If users keep a copy of the conjugate values, then all cases from 4th and 5th bullet point can be remapped to 2nd and 3rd bullet point with some combination of flipping row/col index arrays, using original or conjugated values, and using `trans`/`conjtrans` operations.

Now, the biggest problem of requiring COO to be sorted by row like cuSPARSE, or by columns, is that it almost certainly defeats the purpose of using COO arrays, because users would rather then just use CSR/CSC formats which almost always has a smaller memory footprint (except only when `nrows+1>nnz` for CSR/`ncols+1>nnz` for CSC, which would be rare situations). With that rationale, I think `unsorted` COO being the default is the natural way to go rather than cuSPARSE's approach. Users would need to explicitly sort their arrays and set the `sorted` property to be able to use cuSPARSE backend, otherwise it would throw an `unimplemented` exception saying that cuSPARSE backend does not support unsorted COO matrices and that the matrix property must be set.

The cleanest solution I currently see IMHO is the following:
We decide as a Spec to only support the first three bullet points as otherwise, that bloats up implementations. (Worst case, users need an extra conjugate array to get complete support for 4th and 5th bullet points.) Then:
* The default of not setting the property is associated with being `unsorted`, which matches the CSR/future formats,
* Setting `sorted` property can be associated with the 3rd bullet point, matching the CSR-style `sorted` state
* For the second bullet point, we can add some other `enum`. Some options (and we can't name it `sorted_by_row` alone here):
    * `partially_sorted` -- I would prefer this terse one assuming we don't support 4th/5th bullet points
    * `sorted_by_row_but_not_by_column`
    * `sorted_by_row_with_unsorted_columns`
    * `unsorted_columns_in_sorted_rows`

cuSPARSE users would always need to set at least the `partially_sorted` property, if not the `sorted` property, otherwise oneMKL Interfaces (which would have to track this in its matrix handle for cuSPARSE backend) would throw an `unimplemented` exception. This also allows the oneMKL backend to not leave any performance on the table if the user is comparing performance between cuSPARSE, rocSPARSE, or some other backend versus oneMKL backend.

Reply:
If we were to consider the default of not setting any property as the cuSPARSE style sorted by rows but unsorted columns within the rows, then (now that I think of it,) we can still argue it matches CSR-style default, and `sorted` could still be mapped to the 3rd bullet point. Then we would have instead to introduce a `fully_unsorted_coo` or equivalent `enum` for the Intel oneMKL backend that the users must call and oneMKL Interfaces must keep track of.

Unfortunately, the Intel oneMKL backend has [not currently defined](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-2/sparse-storage-formats.html) what is the default state of not setting the `sorted` property means for the COO format. I can bring this up with the team and present both the options (this comment and previous comment) but unfortunately this will take us at least a week in the best case scenario.

Good thing about this is that it (a) aligns defaults of other libraries, (b) still leaves room for supporting fully unsorted COO matrices with `oneMKL` backends, and (c) **IF** Intel oneMKL backed decides to match the choice of the Spec, then it does not have Intel oneMKL backend leaving performance on the table.

Regardless of the choice, it would be important for both oneMKL Interfaces and Intel oneMKL backend to match each others' approach, otherwise there would be some tricky headaches in the backend implementations.

FYI: @noffermans, @baeseung-intel, @spencerpatty.

Reply:
Do we have a conclusion on this topic @gajanan-choudhary? I'm fine with introducing the `partially_sorted` property in the spec and implementation if needed.

Reply:
Lets keep it to a simple set to begin with and then maybe allow it to grow over time:  For now as I understand it,  cuSPARSE uses "sorted == sorted_like_csr"  for COO.   Anything more than that is extra and non-standard in the libraries at this point.  oneMKL Sparse BLAS does not yet have support for COO sortedness, but will likely start  there...  Maybe we use "sorted_by_rows"  and "sorted_by_columns"  with "sorted = sorted_by_rows" by default ...

Reply:
I don't think `sorted_by_rows` or `sorted_by_columns` is suited for CSR because this format can only be sorted in one way. I would much rather keep the `sorted` name and allow it for CSR and COO formats.
Adding `sorted_by_row` for the COO format is fine with me. To me that name already implies that the columns do not need to be sorted, and it will be clarified in the spec.
I can start working on that soon.

Reply:
hmm,  I think if we use a term like `sorted_by_row` then it should actually mean sorted by rows and subordinate to rows, the column ids are sorted as well,  which is what I think cuSparse means by it.

Because of this ambiguity, we might as well then just have a single `sorted` option and for COO it means sorted by row and then subordinate by columns within rows ...  and for CSR or CSC it means the natural way ...  such ambiguity is unfortunate.  If someone wants COO sorted by column major approach, then they could create a matrix handle with the transpose (swap ia/ja) and then sort it ...

Reply:
I think there is a misunderstanding, cuSPARSE allows the column indices to be unsorted, from their [doc](https://docs.nvidia.com/cuda/cusparse/index.html#coordinate-coo):
```quote
cuSPARSE supports both sorted and unsorted column indices within a given row.
```
To me `sorted_by_row` should not imply anything on the order of the columns indices within a row.

Reply:
oh you are right.  So sorted and unsorted column indices, but always sorted by rows ...   so in your mind there are three states:

- **unsorted** - completely free row/column index  (unsupported by cuSPARSE, very hard to deal with in general, Intel oneMKL has impl of SpMV for this, but perf is not good)
- **sorted_by_rows** - sorted by row, but no guarantees about sortedness of column indicies within the row (cuSPARSE supports,  Intel oneMKL doesn't currently have this as an option, but it would greatly help with basic SpMV impls
- **sorted**  - completely sorted by rows and sorted by column indices within a given row

maybe names can be adjusted slightly, but I guess these the states you are suggesting ?  I don't know if we really want to support the fully unsorted case... It prevents a lot of algorithm improvements and often requires either massive temporary workspaces or use of atomics, both of which can be expensive and less than performant ...

Reply:
Fine with me,  maybe we then have two property flags that could be applied by users:  (we only have CSR/COO formats supported now, but looking forward to potential extensions)

- `sorted_by_rows`  -- row indices are grouped, but no guarantees on column indices within rows
    - CSR - same as unsorted state
    - CSC - invalid argument, unsupported
    - COO - supported
    - BSR - same as unsorted state
    - SELL - same as unsorted state (before ELLPACK like conversion from compressed rows to blocks with backfill and column major format within blocks, the column indices have no order within rows)
    - DIA - invalid argument, unsupported (diagonal format is sorted naturally by diagonals)
- `sorted`  -- row/col indices are grouped in fully sorted state according to the following properties
    - CSR - column indices are sorted within rows
    - CSC - row indices are sorted within columns 
    - COO - column indices are sorted within rows
    - BSR - column indices are sorted within rows
    - SELL - before ELLPACK like conversion from compressed rows to blocks with backfill and column major format within blocks, the column indices are sorted within rows
    - DIA - diagonal format is naturally sorted by diagonals already, so there is no unsorted diagonal representation


Any other formats worth trying to add to list to make sure these are sufficient for future ?

Reply:
maybe we want to change `sorted_by_rows` to something like `sorted_by_rows_only`  ?  so it is more clear that it is a partial sorted state ?

Reply:
So I see we are leaning toward 3 states and are clear about what `sorted` state is for all formats. What should be the default unsorted state for COO format, though? Accordingly, that enum can be removed and the remaining one can be kept around. That circles back to my comment:
1. If the default expectation in the Spec is for COO to be `partially_sorted`/`sorted_by_rows`, then that enum can be removed, and simply an `unsorted` enum can be introduced for users who want to deal with fully unsorted COO formats.
1. If the default expectation in the Spec is for COO to be `unsorted`, then that enum can be removed, and we must introduce a `partially_sorted`/`sorted_by_rows` enum.

I would prefer option (1). It is cleaner and may deliver balanced performance between fully unsorted and fully sorted COO formats, and users have to make an explicit choice that they want to use fully unsorted situation knowing that the perf may not be great (which the spec/docs will clarify).

If we disallow CSC-style sorted states of COO format, then thinking about a future `sort_matrix` API for these two cases, the one that Intel oneMKL currently has, `sycl::event sort_matrix(sycl::queue &queue, matrix_handle_t spMat, const std::vector<sycl::event> &dependencies = {})`, will then always sort to `sorted` property state (no partial sorting allowed), i.e, sorted rows and sorted column indices within rows for COO format.

Reply:
If we go down `enum property {unsorted, sorted, ...}`, however, then Intel oneMKL must update its documentation of what is expected default state of COO format and accordingly add/match that `enum`.

Right now, the documentation does not talk about that. There is the [sparse storage formats page](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-2/sparse-storage-formats.html) that doesn't tell what the default is, and the [`sort_matrix` API](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-2/oneapi-mkl-sparse-sort-matrix.html) does not support COO format yet because of the ambiguities discussed here. There is the [`set_coo_data` API page](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-2/oneapi-mkl-sparse-set-coo-data.html) that also does not mention what is the expected default COO state.

Reply:
@Rbiessy, you mentioned in your comment:
> It turns out rocSPARSE also support unsorted COO format so it would be unfortunate to disallow this I think.

Where do you see that? In the [rocSPARSE user guide](https://rocm.docs.amd.com/projects/rocSPARSE/en/latest/how-to/basics.html#coo-storage-format), I see:
> The COO matrix is expected to be sorted by row indices and column indices per row. Furthermore, each pair of indices should appear only once.

Reply:
Thanks for the suggestions. I've created the spec PR: https://github.com/uxlfoundation/oneAPI-spec/pull/583 and applied the changes in https://github.com/oneapi-src/oneMKL/pull/527/commits/d23b24d0abee93518b0df6cb6af13ac21792f77d. The tests now always set the properties needed by a backend.

I overall agree with your last comment Spencer. I suggest to specify that using `sorted_by_rows` for CSR for instance will throw an exception instead of saying that it does the same as unsorted. I don't think the `_only` suffix is needed.

Gajanan, I don't get why you're saying changing the default of the COO format to be partially sorted is cleaner. The idea of the properties is that users must set them to provide a guarantee on the matrix, it seems natural that the default state has no properties and is thus unsorted. The oneMKL Interface spec already specifies that [here](https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneMKL/source/domains/spblas/data_types/format-descriptions.rst?plain=1#L29), I don't think we can spend time to change this decision.
You're right about rocSPARSE, I looked at the tests and misremembered what I did. This backend always need to have the `sorted` property for COO and CSR. I understand the oneMKL product may not support unsorted COO matrices in the future but I would rather keep it supported in oneMKL Interface as long as the product does.

Reply:
> The idea of the properties is that users must set them to provide a guarantee on the matrix, it seems natural that the default state has no properties and is thus unsorted.

I completely agree with that. However, my reasons for saying that _"keeping partial sortedness as default requirement is cleaner"_ are:
1. That makes users' code functionally portable as well as performance-portable as-is without code changes across NVIDIA and Intel in both directions. If partially-sorted is _not_ the default, then users would need to deal with writing code for sorting (difficult) or for setting the matrix property (easy). At the same time, from performance portability perspective, people going from NVidia backend to oneMKL backend will find that the performance of oneMKL backend is far worse than expected unless they add in the easily forgettable one-liner to set the partially sorted/sorted-by-rows `property`.
1. Users have to make an explicit choice through `set_matrix_property` call to use fully unsorted COO matrices so that it doesn't come as a surprise to them that the performance is not great (which the Spec/docs must clarify).
1. Although we do want users to call `set_matrix_property` whenever possible, it is easy for users to miss out doing that mainly because almost none of our own code examples contain `set_matrix_property` calls for users to be introduced to it or be familiar with it (which is our fault, of course!).
1. Adding restrictions (partially sorted) first and relaxing them later (fully unsorted) is cleaner/less problematic than having a completely relaxed API and later being forced to add restrictions leading to broken backward compatibility and broken user code.
1. (Minor reason) we don't have to worry about any ambiguity in naming the `unsorted` enum property that otherwise exists in partially sorted case. On the other hand, if `unsorted` is the default, then I don't see how either `partially_sorted`/`sorted_by_rows`/any other _short_ names are clear without looking up the documentation (which is not a big deal, I suppose). The names are open to interpretation until users hit a problem and then go look up the docs.

> The oneMKL Interface spec already specifies that [here](https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneMKL/source/domains/spblas/data_types/format-descriptions.rst?plain=1#L29), I don't think we can spend time to change this decision.

We haven't formally released a new version of oneMKL specification since our sparse BLAS spec change in https://github.com/uxlfoundation/oneAPI-spec/pull/522, so this change really isn't a big deal. Only the discussions here and getting to an agreement may take some time, that's all. Although Intel oneMKL supports fully unsorted COO matrices, that is not necessarily a reason why the oneMKL Specification and therefore this repo too should support fully unsorted COO matrices.
_______________________________________
Having said all that, I am still not saying we should go either for `unsorted` or for `sorted_by_rows` as default. I am just saying that the discussion must continue and we must present good reasons to select one over the other. What advantages do you see of going down the route of keeping `unsorted` as default and introducing `sorted_by_rows` enum with regards to the oneMKL Spec and various backends?

Reply:
1. Note that users transitioning from Nvidia to oneMKL Interface will get an `unimplemented` exception, with a clear explanation of the fix, if the COO format is used unsorted so there is no risk of getting bad performance in this case. I agree that having `sorted_by_rows` as the default would make the library more portable without code changes but if that is an important factor having `sorted` as the default would also make it portable with rocSPARSE. The way I see it, any users writing a performant application will use sorted data and will easily see that they need to set the property when testing on Nvidia HW.
2. We can make the same argument for keeping `unsorted` as the default. Users need to make an explicit choice for setting the `sorted_by_rows` property so it does not come as a surprise if the data is not sorted by rows and the results are incorrect. This is a stronger argument in favor of keeping `unsorted` the default IMO.
3. We could change the example using mklcpu+cusparse to use the COO format with `sorted_by_rows` property if that helps. This will make the example more complex though.

I agree on points 4 and 5. I just find the arguments for keeping `unsorted` the default stronger.

I am concerned with the time as we are aiming to merge this and the rocSPARSE PR by the end of October. My point is that this is only achievable if we stick to the decisions that we took when writing the specification. I don't think this discussion is worth missing the deadline.

Reply:
Fair points, @Rbiessy. I realize that my argument (4) also has one potential issue down the line: If we decide to keep partially sorted as default for now (without introducing an `unsorted` enum with the intent of keeping our options open), but later pivot to default being fully unsorted and introducing partial sort enums instead, then although users' code would remain functional, they would see silent perf regressions without modify their code to call `set_matrix_property`. That's also not something we want, so (4) doesn't really let us keep our options open in the future as I earlier thought it should. It becomes same as making the decision _now_.

I'm fine with keeping fully unsorted COO as the default state and keeping {`sorted_by_rows`, `sorted`} `property` enums in the oneAPI Spec and consequently in this repo.

Reply:
> We could change the example using mklcpu+cusparse to use the COO format with sorted_by_rows property if that helps. This will make the example more complex though.

Please also let's do this. We want users to call `set_matrix_property` whenever it is applicable, and examples are a good place to encourage that.

Reply:
Great! I have updated the example in https://github.com/oneapi-src/oneMKL/pull/527/commits/18156e5f1a152cbcfa0fd567e76625286a2e08c3. Turns out there were some issues with that example as it is only compiled if MKLCPU and cuSPARSE backends are enabled!
This log show that test being run with the recent changes: [test_log_mklcpu_cusparse.txt](https://github.com/user-attachments/files/17151329/test_log_mklcpu_cusparse.txt)